Add quantized op support to llama runner

larryliu0820 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #3062
* #3055

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D56197863](https://our.internmc.facebook.com/intern/diff/D56197863)

larryliu0820 committed:
Add quantized op support to llama runner

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

larryliu0820 committed:
Update on "Add quantized op support to llama runner"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

larryliu0820 commented:
@larryliu0820 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D56197863).

larryliu0820 committed:
Update on "Add quantized op support to llama runner"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D56197863](https://our.internmc.facebook.com/intern/diff/D56197863)

[ghstack-poisoned]

larryliu0820 committed:
Update on "Add quantized op support to llama runner"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D56197863](https://our.internmc.facebook.com/intern/diff/D56197863)

[ghstack-poisoned]

larryliu0820 committed:
Update on "Add quantized op support to llama runner"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D56197863](https://our.internmc.facebook.com/intern/diff/D56197863)

[ghstack-poisoned]

larryliu0820 committed:
Update on "Add quantized op support to llama runner"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D56197863](https://our.internmc.facebook.com/intern/diff/D56197863)

[ghstack-poisoned]

kimishpatel commented on .ci/scripts/test_llama.sh:
```diff
@@ -84,7 +91,7 @@ cmake_build_llama_runner() {
         -DEXECUTORCH_BUILD_OPTIMIZED=ON \
         -DEXECUTORCH_BUILD_XNNPACK="$XNNPACK" \
-        -DEXECUTORCH_BUILD_OPTIMIZED=ON \
+        -DEXECUTORCH_BUILD_QUANTIZED="$QE" \
```
Can we not always build? haveing so many options for build feels like additional burden for users. Maybe do default opt-in?

mikekgfb reviewed (approved):
Thanks!

kimishpatel commented on .ci/scripts/test_llama.sh:
```diff
@@ -126,9 +133,15 @@ fi
-  EXPORT_ARGS="${EXPORT_ARGS} -kv --use_sdpa_with_kv_cache -X -qmode 8da4w -G 128"
+EXPORT_ARGS="-c stories110M.pt -p ${PARAMS} -d ${DTYPE} -n ${EXPORTED_MODEL_NAME} -kv"
+if [[ "${XNNPACK}" == "ON" ]]; then
+  EXPORT_ARGS="${EXPORT_ARGS} -X -qmode 8da4w -G 128"
```
nit: does += operator work?

kimishpatel commented on .ci/scripts/test_quantized_aot_lib.sh:
```diff
@@ -24,7 +24,7 @@ build_cmake_quantized_aot_lib() {
       -DCMAKE_BUILD_TYPE=Release \
       -DCMAKE_PREFIX_PATH="$CMAKE_PREFIX_PATH" \
-      -DEXECUTORCH_BUILD_QUANTIZED=ON \
+      -DEXECUTORCH_BUILD_QUANTIZED_OPS_AOT=ON \
```
same as before. default ON would be good

kimishpatel commented on examples/models/llama2/CMakeLists.txt:
```diff
@@ -91,6 +92,7 @@ add_subdirectory(runner)
 if(EXECUTORCH_USE_TIKTOKEN)
   # find RE2 for tokenizer
   set(ABSL_ENABLE_INSTALL ON)
+  set(ABSL_PROPAGATE_CXX_STD ON)
```
oh we depend on abseil for tiktoken?

larryliu0820 replied:
Yeah, tiktoken -> re2 -> abseil

lucylq commented on examples/models/llama2/CMakeLists.txt:
```diff
@@ -91,6 +92,7 @@ add_subdirectory(runner)
 if(EXECUTORCH_USE_TIKTOKEN)
   # find RE2 for tokenizer
   set(ABSL_ENABLE_INSTALL ON)
+  set(ABSL_PROPAGATE_CXX_STD ON)
```
no tests using this path yet right?

larryliu0820 replied:
Not yet

lucylq commented on .ci/scripts/test_llama.sh:
```diff
@@ -126,9 +133,15 @@ fi
+  EXPORT_ARGS="${EXPORT_ARGS} --use_sdpa_with_kv_cache"
+fi
+if [[ "${QE}" == "ON" ]]; then
+  EXPORT_ARGS="${EXPORT_ARGS} --embedding-quantize 8,1024"
```
thanks for adding tests!

larryliu0820 committed:
Update on "Add quantized op support to llama runner"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D56197863](https://our.internmc.facebook.com/intern/diff/D56197863)

[ghstack-poisoned]

larryliu0820 commented:
@larryliu0820 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D56197863).
